File: linkedlist/leetcode_Remove_Nth_Node_From_End_of_List.java
/*Remove_Nth_Node_From_End_of_List

Given a linked list, remove the nth node from the end of list and return its head.
For example,
Given linked list: 1->2->3->4->5, and n = 2.
After removing the second node from the end, the linked list becomes 1->2->3->5.
Note:
Given n will always be valid.
Try to do this in one pass.
*/


class leetcode_Remove_Nth_Node_From_End_of_List {
    public static void main(String[] args) {
    }
    public static ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null) {
            return head;
        }
        ListNode res = new ListNode(0);
        res.next = head;
        ListNode tail = res;
        ListNode node = res;
        while (n > 0) {
            // Create a head node and tail node with length between them is n.
            tail = tail.next; // Assume length of input linkedlist is no less than n.
            --n;
        }
        while (tail.next != null) {
            // when tail reaches end (null), head is the node to removed.
            tail = tail.next;
            node = node.next;
        }
        node.next = node.next.next;
        return res.next;        
    }
}


class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

File: linkedlist/leetcode_Remove_Nth_Node_From_End_of_List.cpp
/*
 * Copyright 2015
 * Author: madarfacar
 *
 * leetcode remove nth node from end of list
 */

#include <stdio.h>


struct ListNode {
  int val;
  ListNode *next;
  explicit ListNode(int x) : val(x), next(NULL) {}
};


void printLinkedList(ListNode* head) {
  while (head != NULL) {
    printf("%d -> ", head->val);
    head = head->next;
  }
  printf("NULL\n\n");
}


class Solution {
 public:
  ListNode* removeNthFromEnd(ListNode* head, int n) {
    if (head == NULL) {
      return head;
    }

    ListNode* end = head;
    bool largerThanLength = false;

    // Move pointer to the position that the distance between head and this ptr
    // is given length. Move with best effort to handle the case that n might
    // be larger than the lenght of this linkedlist.
    while (!largerThanLength && n > 1) {
      end = end->next;
      largerThanLength = end == NULL;
      --n;
    }

    if (largerThanLength) {
      // If n is larger than length, do nothing.
      return head;
    }

    // Create dummy node to attach to the beginning.
    ListNode* dummy = new ListNode(0);
    ListNode* start = dummy;
    dummy->next = head;

    // Move end to the last ListNode.
    while (end->next != NULL) {
      end = end->next;
      start = start->next;
    }

    // Unlink start->next;
    start->next = start->next->next;
    // Well since the destructor of ListNode is undefined, it does nothing.
    delete start;

    return dummy->next;
  }
};


int main() {
  Solution sol;

  ListNode* head = new ListNode(1);
  ListNode* next = new ListNode(2);
  ListNode* nnext = new ListNode(3);
  head->next = next;
  next->next = nnext;

  printf("Before:\n");
  printLinkedList(head);

  ListNode*res = sol.removeNthFromEnd(head, 4);
  printf("After:\n");
  printLinkedList(res);
}
